perma_web/warc_server/pywb_config.py

```diff
@@ -17,58 +15,149 @@
 from django.core.exceptions import DisallowedHost
 from django.core.cache import cache as django_cache
 
+from pywb.cdx.cdxserver import CDXServer
 from pywb.cdx.cdxsource import CDXSource
 from pywb.framework import archivalrouter
 from pywb.framework.wbrequestresponse import WbResponse
+from pywb.framework.memento import MementoResponse, LINK_FORMAT
 from pywb.rewrite.wburl import WbUrl
 from pywb.utils.loaders import BlockLoader, LimitReader
-from pywb.warc.cdxindexer import write_cdx_index
 from pywb.webapp.handlers import WBHandler
 from pywb.webapp.query_handler import QueryHandler
 from pywb.webapp.pywb_init import create_wb_handler
 from pywb.webapp.views import add_env_globals
+from pywb.webapp.pywb_init import create_wb_router
+from pywb.utils.wbexception import NotFoundException
 
-from perma.models import Link
+from perma.models import CDXLine, Asset, Link
+
+# Assumes post November 2013 GUID format
+GUID_REGEX = r'([a-zA-Z0-9]+(-[a-zA-Z0-9]+)+)'
 
 
 # include guid in CDX requests
-class Route(archivalrouter.Route):
+class PermaRoute(archivalrouter.Route):
     def apply_filters(self, wbrequest, matcher):
-        wbrequest.custom_params['guid'] = matcher.group(1)
+        """Parse the GUID and find the CDXLine in the DB"""
+
+        guid = matcher.group(1)
+        urlkey = surt(wbrequest.wb_url_str)
+
+        try:
+            # This will filter out links that have user_deleted=True
+            link = Link.objects.get(guid=guid)
+        except Link.DoesNotExist:
+            raise NotFoundException()
+
+        try:
+            line = CDXLine.objects.get(urlkey=urlkey,
```

Comment:
This query and the custom_params cache need to handle multiple lines with the same urlkey. For example, here are the first two CDX lines for a capture of "google.com", which redirects to "www.google.com":

```
com,google)/ 20150423160350 http://google.com/ text/html 301 c79f5572f672361bc097676cb5da9d4aa956c8b9 - - 712 1673 2015/4/23/16/3/BY3G-VVCQ/archive.warc.gz
com,google)/ 20150423160350 http://www.google.com/ text/html 302 43c3c326039fde13b0473b08d729e6fcd7c643b3 - - 1012 2960 2015/4/23/16/3/BY3G-VVCQ/archive.warc.gz
```

Note that the surt for both lines is the same because the www is normalized away. In a search for `com,google)/` we need to return *both* lines -- then pywb will detect the self-redirect in the first line, discard it, and move on to the second one. Strange but true.